## Abstract {.page_break_before}
Analyzing genomics data, including RNA-, ATAC- and ChIP-sequencing, requires multiple types of support data such as genome sequence and gene annotations.
Many of these resources can be retrieved from different organizations, exist in multiple versions and may be generated by different methods.
What datasets to use depends on the context of the research, such as collaboration partners, data reuse or dataset quality.
As such, making an informed decision is essential. While many analysis pipelines are available, these mostly require manual downloading and management of genome-related resources.
This can be tedious and error-prone and does not allow for full computational reproducibility.

Here we present genomepy, a command line tool and Python API that can navigate the genome databases of Ensembl, UCSC and NCBI.
Genomepy can search and install genome sequences and gene annotation data from these providers in a consistent, reproducible and documented manner.
The search function retrieves genomes related to the search term for all of the available providers, allowing the user to make an informed decision.
The installation of the genome can be modified to obtain a soft-, hard- or unmasked version, or specific chromosomes or scaffolds filtered by regular expressions.
Gene annotation data can be downloaded in addition to the genome sequence, which genomepy checks for compatibility with other bioinformatic tools.
Finally, genomepy can automatically create indexes for commonly used aligners.

To summarize, genomepy is a straightforward tool to find, download and index genomes.
It can be used to obtain genomes, gene annotations and additional support files in a consistent and automatic fashion.
Genomepy is freely available at https://github.com/vanheeringen-lab/genomepy and can be installed using Bioconda and Pip.
