## Abstract {.page_break_before}

### Summary
Analyzing genomics data, including RNA-, ATAC- and ChIP-sequencing, requires multiple types of support data, such as genome sequence and gene annotations.
These resources can generally be retrieved from multiple organizations, where they exist at multiple versions, and may have been generated with varying methods.
Which data to use depends on the context of the research, such as collaboration partners, data reuse or data quality.
Many of the bioinformatic workflows and pipelines available to date require the user to make this informed decision and supply the support data.
Obtaining this data can be a tedious and error-prone process and does not allow for full computational reproducibility.

Here we present genomepy, a quality-of-life enhancement tool, that can navigate the genome databases of Ensembl, UCSC and NCBI.
Genomepy can search and install genome sequences and gene annotation data from these providers in a consistent, reproducible and documented manner.
The search function retrieves all genomes related to the search term, and can do so for each provider to allow the user to make an informed decision.
The install function downloads the genome (and optionally the gene annotations) with sensible defaults, while providing full control to advanced features.
Genomepy can optionally create genome indexes for commonly used aligners.

Genomepy provides this functionality and more via command line interface and Python application programming interface to allow for integration in automated pipelines.

### Availability and implementation
Genomepy can be installed using [Bioconda](https://anaconda.org/bioconda/genomepy) and [Pip](https://pypi.org/project/genomepy/), and is available at https://github.com/vanheeringen-lab/genomepy.
