{##

    To create the webpage locally, run `build/run_on_localhost.sh` and clock the http link.
    
    Aim: up to 2 pages; this is approx. 1,300 words or 1,000 words plus one figure.
##}

## Abstract {.page_break_before}

Analyzing a functional genomics experiment, such as ATAC-, ChIP- or RNA-sequencing, requires reference data including a genome assembly and gene annotation.
These resources can generally be retrieved from different organizations and in different versions.
Most bioinformatic workflows require the user to supply this genomic data manually, which can be a tedious and error-prone process.

Here we present genomepy, which can search, download, and preprocess the right genomic data for your analysis.
Genomepy can search genomic data on NCBI, Ensembl, UCSC and GENCODE, and compare available gene annotations to enable an informed decision.
The selected genome and gene annotation can be downloaded and preprocessed with sensible, yet controllable, defaults.
Additional supporting data can be automatically generated or downloaded, such as aligner indexes, genome metadata and blacklists.
These functionalities are available from a command line interface, aimed at ease of use and integration in automated pipelines. 
Extended functionality is accessible using a Python application programming interface. 
