{##

    To create the webpage locally, run `build/run_on_localhost.sh` and clock the http link.
    
    Aim: up to 2 pages; this is approx. 1,300 words or 1,000 words plus one figure.
##}

## Abstract {.page_break_before}
Analyzing functional genomics data, including ATAC-, ChIP- and RNA-sequencing, requires genomic data such as a genome assembly and gene annotations.
These resources can generally be retrieved from multiple organizations, at multiple versions, and generated with varying methods.
Meanwhile, most bioinformatic workflows and pipelines require the user to supply this genomic data manually, which can be a tedious and error-prone process.

Here we present genomepy, which can search, download, and preprocess the right genomic data for your analysis.
Genomepy can search genomic data on GENCODE, Ensembl, UCSC and NCBI, and compare available gene annotations, to allow for an informed decision.
The selected genome and gene annotation can be downloaded (from anywhere) and preprocessed with sensible, yet controllable, defaults.
Additional supporting data can be automatically generated, such as aligner indexes, genome metadata and blacklists.
These functionalities are available on command line interface, aimed at ease of use and integration in automated pipelines, 
with extended functionality on the Python application programming interface.
