## Introduction
Its big, its vague, and you just want to run your pipeline yesterday, right?

TODO: figures making it clear these are the major three providers

| Database | Fun aspect |
|:--|:--|
| Ensembl | Generally seen as standard, updates infrequently, incompatible chromosome names |
| UCSC    | multiple GTF formats, GTFs labelled incorrectly |
| NCBI    | Different pipelines, looks like Ensembl, updates frequently |

Table: Genome providers. {#tbl:providers-id}

## Related Work
Many genome providers support downloading from their databases via a variety of different methods, including web portals, FTP archives and REST APIs.
External tools have been developed for some of these to extend their functionality, such as the ncbi-genome-download tool @https://github.com/kblin/ncbi-genome-download, and packages such as ucsc-genomes-downloader @https://pypi.org/project/ucsc-genomes-downloader/ (Python) and metaseqR @https://doi.org/10.1093/nar/gku1273 (R).
However, to the best of our knowledge no tool exists that can work with all three major genome providers.
Once the genome and gene annotations have been obtained, there exists a number of tools that can be used to store these files reproducibly, such as refgenie @https://doi.org/10.1093/gigascience/giz149 and tximeta @https://doi.org/10.1371/journal.pcbi.1007664

We conclude that there is a need for a tool that can provide an overview of the choices available, which genomepy provides, and a tool which can obtain this data simply and reproducibly, which genomepy also provides.
