## Introduction
In order to cope with the explosive increase in Omics data, robust and scalable bioinformatics tools are required.
Many tools have been developed to perform the functions of preprocessing @doi:10.5281/zenodo.3921913, analysis @doi:10.5281/zenodo.824117 and workflow management [@doi:10.12688/f1000research.29032.1; @https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/genomepy.html]
However, many tools cannot obtain all input data automatically, notably the genomic data, which includes the genome assembly, gene annotation and derived files.
These data can be obtained from a variety of different providers, including three major providers, Ensembl @https://doi.org/10.1093/nar/gkz966, UCSC @doi:10.1101/gr.229102 and NCBI @pmc:PMC308837, and many niche providers, such as flybase @https://doi.org/10.1093/nar/gky1003, wormbase @https://doi.org/10.1093/nar/gkz920 or xenbase @doi:10.1093/nar/gkx936.
Each provider has a different method of generating genome assemblies and gene annotations, which can affect available data formats, naming schema, information density, as well as availability, accessibility and relevance.
The differences between these data significantly impact the compatibility of the reference data with research tools @https://doi.org/10.1186/s12864-015-1308-8, other reference databases, and other research.

In order to assist in searching through genome providers for, and standardize the processing of genomic data, we developed genomepy.
The genomepy search function returns all genomes on the three major providers containing the search term in their name, description or accession identifier, as well as genomes matching a taxonomy identifier.
The genomepy install function retrieves a specified genome assembly and related data in a format ready for downstream use.

## Related Work
Ensembl, UCSC and NCBI all support downloading from their individual databases via accessible FTP archives, web portals, and REST APIs.
To access these databases programmatically, there exists several external tools, such as the ncbi-genome-download tool @https://github.com/kblin/ncbi-genome-download and ucsc-genomes-downloader @https://pypi.org/project/ucsc-genomes-downloader.
However, to our knowledge no tool exists that can consistently search or download from all three major genome providers.

There are several existing tools for reproducibly sharing reference data between projects.
Data management tools accept reference data and derived assets such as aligner indexes, and include iGenomes @https://support.illumina.com/sequencing/sequencing_software/igenome.html, refGenie @https://doi.org/10.1093/gigascience/giz149 and Go Get Data @https://doi.org/10.1101/2020.09.10.291377.
These tools excel in their ability to reproducibly share data, a feature which is not present in genomepy, and can be used to obtain and manage previously generated data with ease.
However, these tools require the user to supply the reference data to any new assembly, newer version, or certain assets.

We conclude that there was a need for a tool that can programmatically obtain and preprocess genomic data, which is how genomepy came to be.

![niche for genomepy in the bioinformatics ecosystem.](images/compare.png){#fig:compare height="500px" width="500"px}
