## Introduction
High-throughput sequencing is common practice (since/data volumes)
automation required to process the large volumes consistently
This is possible because of similarities in the sequencing methods and standardized formats
Aligners map the read data to a reference genome or transcriptome.
Both aligners and other steps in the various sequencing data analyses require as input a reference genome and gene annotation
three major genome providers: Ensembl, UCSC and NCBI
Ensembl:
UCSC:
NCBI:
In addition, there are many species specific providers, such as flybase, wormbase, xenbase.

Each provider has a separate method of generating their reference genomes and gene annotations, which can affect data format, naming and density, as well as database versioning and update frequency.
These differences impact the compatibility of the reference data with research tools @https://doi.org/10.1186/s12864-015-1308-8, other reference databases and other research.
Therefore, the choice of reference data is significant importance.
To make this decision on an informed bases, you need an overview of the options.
To get this overview, one could check each website, download the desired data, process and log these steps manually.
For the sake of sanity and reproducibility, it would be better if that could be done in a standardized system.

Here we present genomepy, which can do that.
Search one or all 3 providers
Install genome and gene annotations
Automatic preparation for aligners (genome indexing with pyfaidx @todo, generating support files (chromosome sizes and gaps), matching chromosome names between genome and gene annotation and optional aligner index generation)
Automatic logging for reproducibility
CLI and Python API can be used to automate this step in workflows.

## Related Work
Ensembl, UCSC and NCBI support downloading from their databases via accessible FTP archives, web portals, and REST APIs.
External tools have been developed to programmatically download from one or several databases, such as the ncbi-genome-download tool @https://github.com/kblin/ncbi-genome-download, and packages such as ucsc-genomes-downloader @https://pypi.org/project/ucsc-genomes-downloader/ (Python) and metaseqR @https://doi.org/10.1093/nar/gku1273 (R).
However, to our knowledge no tool exists that can search or download from all three major genome providers.

The reference data may not be ready for direct downstream use.
For instance, many assemblies do not contain gene annotations in the correct format for splice-aware aligners.
Furthermore, many gene annotations have contig (chromosomes/scaffolds) names that do not match the names in the reference genome.
Additional processing steps are required to correct these issues.
Tools exist to address some of these issues, but would be more effective when used in conjunction.

We conclude that there is a need for a tool that can provide an overview of the choices of reference data available, can obtain the specified data, and perform the processing required to utilize the data downstream.
Genomepy was created to fit this need, and does so for both automated and human-supervised workflows.
