## genomepy
The core functionalities of genomepy are `search` and `install`.

### search
Search will query a provider (or all three if none is specified) for the given search term.
The search term normalized for case and whitspace, and input types is identified.
For taxonomy identifiers, all assemblies with matching IDs are returned.
For accession identifiers and text terms, all assemblies containing the term in their respective field are returned.
Additionally text terms found in any other descriptive fields are also returned.

### install
When an assembly has been selected, the name can be passed to the install function.
This function downloads the genome assembly with soft masking, unless different masking is specified.
The assembly is then filtered to exclude alternative regions (unless specified otherwise) and the presence or absence of any specified (regex search) term.
The genome is then indexed using pyfaidx @https://doi.org/10.7287/peerj.preprints.970v1, and contig sizes and gaps are stored in separate files.

If specified, genomepy will attempt to download a gene annotation:
genomepy will search the database for a GFF, GTF, BED or (for UCSC only) text format gene annotation.
The annotation is then processed to output a consistent GTF and BED format gene annotation using the UCSC conversion tools @http://hgdownload.cse.ucsc.edu/admin/exe/.
If the genome was downloaded previously, the contig names are checked for compatibility, and matched to those in the genome if required and possible.

### external providers
External provider often contain novel or more recent assemblies of organisms in their specialized field.
These assemblies may be processed similarly by providing the direct link to the genome and/or gene annotation in the genomepy install command and specifying 'url' as provider.

### plugins
Using the `plugin` function, the generation of aligner indexes can be toggled.
The indexes will automatically generate upon the completion of the install function.

### logging
Download sources, data and time, processing steps and requested filters are all logged in a README file which is stored in the same directory, and updated when further processing is performed with genomepy.



### old
search, download, sensible defaults, reproducible, automatable.
about those defaults...

Install via conda, pip or git.

basic steps in CLI

Repeat steps in API
Extended steps, link to seq2science implementation?
