## genomepy
The core functionalities of genomepy are to search, download and prepare genomes and gene annotations.
These functions are split over the `search` and `install` functions.

Search will query the three major providers for a given search term.
Genomepy can search for text terms in genome names or descriptions, taxonomy identifiers and accession numbers, and will automatically detect which.
The search results are returned with available metadata for review.

An assembly name from a major provider can be passed to the install function, along with the name of the provider if the data is available on multiple.
Alternatively, if the assembly originates from another source, the url to the genome can be passed.
Next, the genome assembly is downloaded with the desired sequence masking level [@http://repeatmasker.org; @doi:10.1089/cmb.2006.13.1028].
By default soft masked genomes are downloaded, but unmasked or hard masked can be downloaded (or generated if required) as well.
Reference assemblies often contain alternate sequences to reflect biological diversity.
For the purpose of sequence alignment however, the best results are obtained if there is one reference per nucleotide.
Therefore genomepy filters out alternative regions, unless specified otherwise.
Additionally, regex filters may be passed to either include or exclude contigs (chromosomes, scaffolds, etc.) by name.
Once filtering is performed, genomepy generates commonly used support files.
The genome is indexed using pyfaidx @https://doi.org/10.7287/peerj.preprints.970v1, and contig sizes and contig gap sizes are collected in separate files.

If specified and available, genomepy will download the gene annotation.
Gene annotations are output in the commonly used GTF and BED formats.
Contig names of the genome and gene annotation are checked for compatibility.
Should these mismatch, genomepy will attempt to match the names in the annotations to the genome.

![workflow for `genomepy search` and `genomepy install`.](images/flow.png){#fig:flow height="400px" width="600"px}

Genomepy facilitates optional processing steps via plugins.
These options can be inspected and toggled with the `genomepy plugin` command line function.
The blacklist plugin downloads blacklists by the Kundaje lab @https://doi.org/10.1038/s41598-019-45839-z for the supported genomes.
Other plugins support the generation of aligner indexes, including DNA aligner indexes for Bowtie2 @https://doi.org/10.1038/nmeth.1923, BWA @https://doi.org/10.1093/bioinformatics/btp324, GMAP @https://doi.org/10.1093/bioinformatics/bti310 or Minimap2 @https://doi.org/10.1093/bioinformatics/bty191, and splice-aware aligners such as STAR @https://doi.org/10.1093/bioinformatics/bts635 and HISAT2 @https://doi.org/10.1038/nmeth.3317.

For data provenance and reproducibility, a README file is kept with the timestamp, URLs to the source files, the steps performed, and filtered contigs.
