## genomepy
The core functionality of genomepy is to search, download and prepare genomes and gene annotations.
These functions are split over two command line functions: `genomepy search` and `genomepy install`.

Search will query the three major providers for a given search term, with the option query only one specified provider.
Genomepy can search for text terms present in genome names or descriptions, taxonomy identifiers and accession numbers, and will automatically detect which is used in the search term.
The search results are returned with available metadata for review.

When an assembly has been selected, the name (as returned from the search function) can be passed to the install function.
The files will be downloaded from the major provider that hosts it.
Users can optionally specify a provider if the data is available on multiple.
Alternatively, if the assembly originates from another source, the url to the genome can be passed.
Next, the genome assembly is downloaded with the desired sequence masking level [@http://repeatmasker.org; @doi:10.1089/cmb.2006.13.1028].
By default soft masked genomes (repetitive sequences written in lower case) are downloaded, but unmasked or hard masked (repetitive sequences written as Ns) can be obtained as well.
If the provider does not have the genome at this masking level, genomepy will edit the FASTA to match.

Reference assemblies often contain alternate sequences to reflect biological diversity.
For the purpose of sequence alignment however, the best results are obtained if there is one reference per nucleotide.
Therefore genomepy filters out alternative regions, unless specified otherwise.
Additionally, regex filters may be passed to either include or exclude contigs (chromosomes, scaffolds, etc.) by name.
For instance to filter out (or filter for) chromosomes, unplaced-, unknown- or mitochondrial sequences.

Once processing is performed, genomepy generates several commonly used support files.
The genome is indexed using pyfaidx @https://doi.org/10.7287/peerj.preprints.970v1, and contig sizes and contig gap sizes are stored in separate files.

If specified, genomepy will attempt to download a gene annotation:
If available, the annotation is processed to the commonly used GTF and BED output formats.
Afterward, contig names between the genome and gene annotation are checked for compatibility.
Should these mismatch, genomepy will attempt to match the names in the annotations to those in the genome.

For data provenance and reproducibility, a README file is kept with the output which logs the URLs to the source files, the steps performed, and contigs filtered out.

Genomepy facilitates several optional processing steps via plugins.
Using the command line, these options can be inspected and toggled with the `genomepy plugin` function.

The blacklist plugin downloads blacklists by the Kundaje lab @https://doi.org/10.1038/s41598-019-45839-z for the supported UCSC genomes (and GRCh38).
The other plugins currently all support the generation of aligner indexes.
These include DNA aligner indexes for Bowtie2 @https://doi.org/10.1038/nmeth.1923, BWA @https://doi.org/10.1093/bioinformatics/btp324, GMAP @https://doi.org/10.1093/bioinformatics/bti310 or Minimap2 @https://doi.org/10.1093/bioinformatics/bty191, and splice-aware aligners such as STAR @https://doi.org/10.1093/bioinformatics/bts635 and HISAT2 @https://doi.org/10.1038/nmeth.3317.

Once the install function has run with one or more aligner plugins, the genomic data is ready for alignment and analysis.
