## Conclusion
<!-- A principal step in all science is making informed decisions. -->
<!-- For genomics projects, choosing a genome to serve as reference assembly is no different. -->
Obtaining suitable genomic data is a principal step in any genomics project.
Genomepy was developed to provide a consistent overview of genomes on the three major providers,
and reproducibly download and process genomic data ready for downstream use.
<!-- offers an overview of the three largest genome providers, serving as a catalyst for this step. -->
<!--  -->
<!-- After choosing an assembly, data must be downloaded and processed for compatibility with downstream tools. -->
<!-- Genomepy provides this functionality, which includes downloading, zipping and unzipping, converting, filtering and altering of the data, while providing logging of the steps undertaken. -->
<!-- Even if the required reference data is not available on the three largest genome providers, genomepy can process external data to provide a consistent output. -->
Using genomepy, a project can utilize genomes from any provider and expect consistent output with data from a major provider or otherwise.
This allows genomepy to automate another step in genomic data preprocessing.
Additionally, it facilitates downstream analysis, by setting up paths to the genome FASTA file, providing genome metadata within the Python Genome object and generating support files.
<!--  -->
<!-- While genomepy makes choices during the processing, each of these can be tuned to the specific needs of a project using the CLI or Python API. -->
Combined, these features make genomepy ideal for integration in automated sequencing workflows, as demonstrated in seq2science @doi:10.5281/zenodo.3921913, paving the way for robust and reproducible analysis.

### Future prospects
<!-- As science strives to become more open, genomepy assists by making the discussed steps easier, and make the output more FAIR @doi:10.1038/sdata.2016.18. -->
Further improvements to openness and reproducibility could be achieved by combining or coupling of genomepy with data management tools such as Go Get Data and RefGenie.
Additionally, further improvements could be made in the ability to specify the assembly version, in order to allow annotations of different versions with an assembly.
<!--  -->
<!-- For these situations, data management tools would be an excellent extension to genomepy. -->
<!-- Simon: Move to future outlook => integration / combination with GGD and RefGenie. In this section you want to convince the reader that there is a gap, and that genomepy fills that specific gap! -->
